jcl/src/java.base/share/classes/java/lang/StringBuilder.java

```diff
@@ -2261,7 +2261,43 @@ public StringBuilder append(CharSequence sequence, int start, int end) {
 	if (sequence == null) {
 		return append(String.valueOf(sequence), start, end);
 	} else if (sequence instanceof String) {
-		return append(((String)sequence).substring(start, end));
+		if (start >= 0 && end >= 0 && start <= end && end <= ((String) sequence).lengthInternal()) {
+			int currentLength = lengthInternal();
+			int currentCapacity = capacityInternal();
+			int newLength = lengthInternal() + end - start;
+			int sbBase = currentLength - start;
+			String sequenceString = (String) sequence;
+
+			if (String.enableCompression) {
+				// Check if the StringBuilder is compressed
+				if (count >= 0) {
+					decompress(newLength);
+				}
+
+				if (newLength > currentCapacity) {
+					ensureCapacityImpl(newLength);
+				}
+
+				for (int i = start; i < end; ++i) {
+					value[sbBase + i] = sequenceString.charAtInternal(i);
+				}
```

Comment:
Why can't this be an arraycopy?

Reply:
I'll try out arraycopy and see the perf impact. Marking this PR as WIP

Reply:
Results of the ArrayCopy benchmark are posted below.